File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
import heapq

class MedianFinder:

    def __init__(self):
        self.arr = []
        
    def addNum(self, num: int) -> None:
        self.arr.insert(self.getPos(num), num)

    def findMedian(self) -> float:
        size = len(self.arr)
        return self.arr[math.ceil(size/2)-1] if size&1 else (self.arr[size//2] + self.arr[(size//2)-1]) / 2
    
    def getPos(self, val):
        arr = self.arr
        start = 0
        end = len(arr) - 1
        while start <= end:
            mid = start + (end-start) // 2
            if arr[mid] == val:
                return mid
            elif arr[mid] < val:
                start = mid + 1
            else:
                end = mid - 1
        return start

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.rs
use std::collections::BinaryHeap;

struct MedianFinder {
    f:BinaryHeap<i32>,
    s:BinaryHeap<i32>
}

impl MedianFinder {

    fn new() -> Self {
        Self{f:BinaryHeap::new(), s:BinaryHeap::new()}
    }
    
    fn add_num(&mut self, num: i32) {
        if self.f.len()==0 {self.f.push(num); return;}
        if *self.f.peek().unwrap()<=num {self.s.push(-num);} else {self.f.push(num);}
        if self.s.len()>self.f.len() { self.f.push(-self.s.pop().unwrap());}
        if self.s.len()+1<self.f.len() { self.s.push(-self.f.pop().unwrap());}
    }
    
    fn find_median(&self) -> f64 {
        let mut ans = *self.f.peek().unwrap() as f64;
        if self.f.len()==self.s.len() {ans=(ans-*self.s.peek().unwrap() as f64)/2.0;}
        ans
    }
}